roadmap/implementers-guide/src/node/dispute/votesdb.md

```diff
@@ -0,0 +1,147 @@
+# Votes DB
+
+Responsible module for tracking all kind of dispute votes by all validators
+for a fixed number of sessions.
+
+
+## Remarks
+
+Votes must be persisted. Disputes might happen long after
+there was a dispute, as such just keeping things in memory
+is thus sufficient, as every blip (OOM, bug, maintanance)
+could cause the node to forget the state.
+
+`UnsignedTransaction`s are OK to be used, since the inner
+element `CommittedCanddidateReceipt` is verifiable.
+
+## ToDo
+
+* Missing Definition of Governance Mode.
+* Is a `PoV` enough or should the whole `CommittedCandidateReceipt` be included?
+## IO
+
+Inputs:
+
+* `VotesDbMessage::*`
+
+Outputs:
+
+* `DisputeConclusionMessage::*`
+* `ChainApiMessage::Blacklist`
+
+## Messages
+
+```rust
+enum VotesDbMessage {
+    /// Allow querying all `Hash`es queried by a particular validator.
+    QueryByValidatorAndSessionIndex {
+        identification: (SessionIndex, ValidatorIndex), response: ResponseChannel<Vec<Hash>>
+    },
+    QueryByValidatorId {
+        identification: ValidatorId, response: ResponseChannel<Vec<Hash>>
+    },
+
+    /// Register a vote for a particular dispute
+    RegisterVote(DisputeGossipMessage<H = Hash, N: BlockNumber>),
+}
+```
+
+```rust
+enum DisputeMessage {
+    /// A dispute is detected
+    Detection {
+        /// unique validator indentification
+        identification: (SessionIndex, ValidatorIndex),
+        response: ResponseChannel<Vec<Hash>>,
+    },
+    /// Concluded a dispute with the following resolution
+    Resolution {
+        /// Resolution of the block can either be,
+        /// the block being valid or not.
+        valid: bool,
+    }
+}
+```
+
+## Helper Structs
+
+```rust
+/// Snapshot of the current vote state and recorded votes
+struct Snapshot {
+    /// all entries must be unique
+    pro: Vec<(ValidatorIndex, SessionIndex)>,
+    /// all entries must be unique
+    cons: Vec<(ValidatorIndex, SessionIndex)>,
+}
+```
+
+```rust
+/// Gossip being sent on escalation to all other
+/// relevant validators for the dispute resolution.
+///
+/// Currently information as provided in `BackedCandidate<H=Hash>` with
+/// some extra info.
+struct DisputeGossipMessage<H = Hash, N: BlockNumber> {
+    /// the committed candidate receipt
+    pub candidate: CommittedCandidateReceipt<H>,
+    /// The block number in question
+    /// TODO: not sure if this is needed given, `CandidateDescriptor`
+    /// TODO: already includes sufficient identification info
+    pub block_number: N,
+    /// session index relevant for the block in question
+    pub session_index: SessionIndex,
+    /// the validity votes per validator
+    pub validity_votes: Vec<ValidityAttestation>,
+    /// validator indices for the particular session, which validated the
+    /// disputed block.
+    /// Invariant: is never all zeros/has at least one bit set
+    pub backing_validators: BitVec<bitvec::order::Lsb0, u8>,
+}
```

Comment:
We can extract one backing signature from the `CommittedCandidateReceipt` yes?  I suppose `ValidityAttestation` includes validity signatures.  It's maybe worth having a detachable negative vote signature on this message.

Reply:
`CommittedCandidateReceipt` does not contain an equiv of `ValidatorSignature`, it contains the `hash` of the `PoV` - is that enough?

Reply:
I'm confused by this then:  If this is a gossip message, then why does it have a bitfield?  We need to actually check the signatures by everyone who votes negative.

Is this message meant to somehow finalist or close the dispute after enough votes?  I'd think some of this data needs to be mentioned on chain earlier.

Reply:
I think you are correct, this is two concerns fused together.

Reply:
I see.  Yes, we maybe need three three on-chain transaction types, roughly like InitiateDispute, DisputeVote, and ResolveDispute.  It's tricky because InitiateDispute maybe requires a couple DisputeVotes, so maybe it just contains them, not sure.  ResolveDispute pulls these bitfields from the chain state, not sure if it should restate them or not.  All might work for both remote and local disputes, but some have additional consequences for local.

Reply:
> `ValidityAttestation`

is just a empty shell, I assume you'd expect this to be a `Statement`. I'd say so.


> I see. Yes, we maybe need three three on-chain transaction types, roughly like InitiateDispute, DisputeVote, and ResolveDispute. It's tricky because InitiateDispute maybe requires a couple DisputeVotes, so maybe it just contains them, not sure. ResolveDispute pulls these bitfields from the chain state, not sure if it should restate them or not.

I think there is no difference between `InitiateDispute` and `DisputeVote` is miniscule if we allow the latter to contain multiple. 

Just to clarify: `ResolveDispute` would be message being sent from the `VotesDB` subsystem to whoever is interested, containing all `+` and `-` votes _including_ their `ValidityAttestations` (or `Statements` or whatever `Signed<T>` we put in there).

Reply:
> We can extract one backing signature from the CommittedCandidateReceipt yes? I suppose ValidityAttestation includes validity signatures. It's maybe worth having a detachable negative vote signature on this message.

No, there's only a collator signature in the candidate receipt. We can extract all backing signatures from `BackedCandidate`s that appear on-chain, though, and we should, based on the `CandidateEvents` that we observe.

> I think there is no difference between InitiateDispute and DisputeVote is miniscule if we allow the latter to contain multiple.

Yeah, I'm in favor of dispute initiation and dispute voting being the same type. But as Bernhard says, these don't appear to be network messages but rather inter-subsystem messages.